### ISSUE TITLE
[idea/request] Forgot password opens Cordova

### ISSUE THREAD

**Issue by Kestanous** [meteor-useraccounts/core#240]
I don't currently have a need for this but, it would be nice if the forgot password link could open a native app. The use case for this is as follows:
* User downloads app and signs up / forgets their password 
* User-Accounts detects thats it is running in Cordova
* User-Accounts sends a verify email / reset your password email
* Their are two links. One for the web and one that opens the native app.

**Comment by splendido** [comment]
I know you can check if you're on Cordova with `Meteor.isCordova` (if I'm not wrong)...

Bu I'm not sure you can detect it [here](https://github.com/meteor-useraccounts/core/blob/master/lib/core.js#L526) which is where the link is composed...

We might think about doing some test on this.

Tnx!

**Comment by Kestanous** [comment]
I can confirm Meteor.isCordova is what will be needed.

I am assuming you are using Accounts.sendResetPasswordEmail? Maybe we could use/make a third party package that could overload that function and adds some magic. Not sure yet but I will keep an eye on it as I move into email setup.

**Comment by splendido** [comment]
Yep, I'm using the standard `Accounts` API.

...using external email packages was also requested with #221.
But relying on the `Accounts` API to generate/consume tokens might be more trustable?
...I wouldn't leave this for a custom implementation, conversely I'd put it more cleare on the Docs we're relying on the core API for security aspects like this.

**Comment by Kestanous** [comment]
Hmm I agree with what you saying for the most part. I wish we could request a new password token and send our own emails. Rather then customizing emails though an API.

**Comment by Kestanous** [comment]
@splendido the meteor function that does this is rather basic https://github.com/meteor/meteor/blob/devel/packages/accounts-password/password_server.js#L382. We could very easily create our own reset password function that is effectively still pure meteor core.

**Comment by splendido** [comment]
Yeah, I guess this would be **the way** also if we'd like to try to change the *default email service*.

The only thing is: we must ensure to keep it up to date with possible future changes on the token object set under `user.services.password.reset`

:+1:

**Comment by Kestanous** [comment]
Given that MDG now lets up make feature requests on Github I made the above issue.

**Comment by splendido** [comment]
Well done!

you might also make a bigger jump and prepare a PR! ...or it could be me

...it should be straight-forward to extract from [the part](https://github.com/meteor/meteor/blob/devel/packages/accounts-password/password_server.js#L371-393) for token generation, put it inside the `Accounts.resetPasswordToken` and call it inside the original `Accounts.sendResetPasswordEmail` ;-)

**Comment by Kestanous** [comment]
That was my plan ^_^ going to tackle that and the rest of my open-source projects starting Sunday.

**Comment by c316** [comment]
What about using bit.ly? They have mobile deep linking so you don't even need to change the URL you send, just use the bit.ly API and generate a short life link. That combined with your URL scheme for your app should get the same result. 

I imagine there could be a security concern here, but given the length of the token keys and their short life the risk seems minimal (to me). 

http://blog.bitly.com/post/113868146719/introducing-mobile-deep-linking-by-bitly

**Comment by c316** [comment]
Just looked closer here and noticed that it isn't free...bummer. I'm going to keep looking though.

**Comment by c316** [comment]
This looks promising, what do you think? https://timseverien.com/projects/deep-link/

**Comment by Kestanous** [comment]
:+1: for deep-link.js

**Comment by c316** [comment]
@Kestanous have you implemented a project with deep-link.js? I've been using a different method, but I want to see how well your way seems to be working (if you had used deep-link.js).

**Comment by Kestanous** [comment]
@c316 unfortunately my teem has been busy with other things. We plan on coming back to user account in the next week or two (will be helping with user accounts v2 as we can). This is one of the issues that we are hopping to tackle. Would you mind sharing your solution?

**Comment by c316** [comment]
Sure, but I'm don't have a final solution, but I also plan to work on
completing this next week.

Right now I'm using https://github.com/EddyVerbruggen/Custom-URL-scheme and
NGINX to take the standard URL, something like
https://mydomain.com/verify-email/token and redirect it to
mygreatapp://verify-email/token which then opens the mobile app and
completes the email verification.

I am going to convert this flow to confirm the email via standard
verify-email link, then have the redirect attempt to launch the deeplink
url. If that doesn't work, then I know the app either isn't installed or it
isn't a mobile device. Then I can launch the app store or a notification
page (for non-mobile devices). This is where deep-link.js might work
nicely, but I'm still not sure where in the flow I should be putting it.

I'll post a link here when it is complete.

On Mon, May 4, 2015 at 3:29 PM, Sivli Kestanous <notifications@github.com>
wrote:

> @c316 <https://github.com/c316> unfortunately my teem has been busy with
> other things. We plan on coming back to user account in the next week or
> two (will be helping with user accounts v2 as we can). This is one of the
> issues that we are hopping to tackle. Would you mind sharing your solution?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/meteor-useraccounts/core/issues/240#issuecomment-98842159>
> .
>



-- 
Joshua Bechard
--
Learn to do right! Seek justice, encourage the oppressed.
Defend the cause of the fatherless,
plead the case of the widow.
Isaiah 1:17
--

**Comment by Kestanous** [comment]
@c316 Looks like you made some headway. Thanks for the update! This looks very doable.
